### PR TITLE
[Emulator tests] Add proxy server - 2

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description
- Follow up PR on https://github.com/GoogleCloudPlatform/gcsfuse/pull/2728
- Config.go to parse the config file.
- Added operation manager for retrieve operation and add operation.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Added
3. Integration tests - NA

### CODE REVIEW

File: tools/integration_tests/emulator_tests/proxy_server/config.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+type RetryConfig struct {
+	// The name of the method to apply retries to (e.g., JsonCreate, JsonStat).
+	Method string `yaml:"method"`
+	// Retry instruction (e.g., return-503, stall-33s-after-20K).
+	RetryInstruction string `yaml:"retryInstruction"`
+	// Number of times to retry.
+	RetryCount int `yaml:"retryCount"`
+	// Number of retry attempts to skip.
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/config.go:31]: -> // Number of starting retry attempts to skip.

Comment (Tulsishah) [reply tools/integration_tests/emulator_tests/proxy_server/config.go]: Done

File: tools/integration_tests/emulator_tests/proxy_server/config.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	// Retry instruction (e.g., return-503, stall-33s-after-20K).
+	RetryInstruction string `yaml:"retryInstruction"`
+	// Number of times to retry.
+	RetryCount int `yaml:"retryCount"`
+	// Number of starting retry attempts to skip.
+	SkipCount int `yaml:"skipCount"`
+}
+
+type Config struct {
+	// Target host address for the emulator.
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/config.go:36]: Let's follow the best practices of documentation - https://google.github.io/styleguide/go/best-practices.html#documentation

E.g.
// Target host address for the emulator. ==> // TargetHost is the address of emulator server to which proxy server interacts.

File: tools/integration_tests/emulator_tests/proxy_server/config.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	viper.SetConfigFile(configPath)
+	if err := viper.ReadInConfig(); err != nil {
+		return nil, fmt.Errorf("error reading config file, %s", err)
+	}
+
+	if err := viper.Unmarshal(&config); err != nil {
+		return nil, fmt.Errorf("unable to decode into struct, %v", err)
+	}
+
+	// Access the values from the struct
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/config.go:53]: Let's create a method which prints the value of config, use it here.
Also, we can remove this comment.

File: tools/integration_tests/emulator_tests/proxy_server/config_test.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+package proxy_server
+
+import (
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseConfigFile(t *testing.T) {
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/config_test.go:24]: Can we add test for empty content file?

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+func NewOperationManager(config Config) *OperationManager {
+	rc := make(map[RequestType][]RetryConfig)
+	om := &OperationManager{
+		retryConfigs: rc,
+	}
+	for _, retryConfig := range config.RetryConfig {
+		om.addRetryConfig(retryConfig)
+	}
+	log.Printf("%+v\n", om)
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/operation_manager.go:35]: Is there any way to add these print statement only in debug mode?

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+			configs = configs[1:]
+			om.retryConfigs[requestType] = configs
+		}
+	}
+	return ""
+}
+
+func (om *OperationManager) addRetryConfig(rc RetryConfig) {
+	rt := RequestType(rc.Method)
+	println(rt)
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/operation_manager.go:67]: Let's move all these print statement under debug flags.

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager_test.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	// Assert that retryConfigs are initialized correctly
+	assert.Len(t, om.retryConfigs, 2)
+	assert.Len(t, om.retryConfigs["JsonCreate"], 1)
+	assert.Len(t, om.retryConfigs["JsonStat"], 1)
+
+	assert.Equal(t, "return-503", om.retryConfigs["JsonCreate"][0].RetryInstruction)
+	assert.Equal(t, "stall-33s-after-20K", om.retryConfigs["JsonStat"][0].RetryInstruction)
+}
+
+func TestRetrieveOperation(t *testing.T) {
```
Comment (raj-prince) [inline tools/integration_tests/emulator_tests/proxy_server/operation_manager_test.go:42]: Can we add a test?
- with two retry configs (both same request type, both for different request type)

**Review by raj-prince — COMMENTED** [review]
a couple of minor comments.

File: tools/integration_tests/emulator_tests/proxy_server/config_test.go (hunk 4 above)
Comment (Tulsishah) [reply tools/integration_tests/emulator_tests/proxy_server/config_test.go:24]: Done

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager_test.go (hunk 7 above)
Comment (Tulsishah) [reply tools/integration_tests/emulator_tests/proxy_server/operation_manager_test.go:42]: Done

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager.go (hunk 5 above)
Comment (Tulsishah) [reply tools/integration_tests/emulator_tests/proxy_server/operation_manager.go:35]: I'll add main.go file in upcoming PR, their I'll introduce this flag and cover this comment.

File: tools/integration_tests/emulator_tests/proxy_server/operation_manager.go (hunk 6 above)
Comment (Tulsishah) [reply tools/integration_tests/emulator_tests/proxy_server/operation_manager.go:67]: I'll add main.go file in upcoming PR, their I'll introduce this flag and cover this comment.

**Review by raj-prince — DISMISSED** [review]

**Review by raj-prince — APPROVED** [review]
